### PR TITLE
Adds support to callback in breadcrumbs and HTMLElement as target panel for place requests.

### PR DESCRIPTION
**Pull request by paulovmr**

### CODE REVIEW

File: uberfire-api/src/main/java/org/uberfire/mvp/impl/PathPlaceRequest.java [hunk 1]
```diff
@@ -124,6 +124,7 @@ public boolean equals( Object o ) {
     public int hashCode() {
         int result = super.hashCode();
```
Comment (ederign) [inline uberfire-api/src/main/java/org/uberfire/mvp/impl/PathPlaceRequest.java]: super.hashcode has result = ~~result call?

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbs.java [hunk 2]
```diff
@@ -117,9 +121,54 @@ public void clearBreadCrumbsAndToolBars( String associatedPerspective ) {
...
+     *
+     * @param associatedPerspective  perspective associated with the breadcrumb
+     * @param breadCrumbLabel        label of the breadcrumb
+     * @param associatedPlaceRequest place request associated with the breadcrumb
+     * @param addTo                  target content panel of the place request
+     */
+    public void addBreadCrumb( final String associatedPerspective,
+                               final String breadCrumbLabel,
+                               final DefaultPlaceRequest associatedPlaceRequest,
+                               final HasWidgets addTo ) {
```
Comment (ederign) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbs.java]: Why use here HasWidgets and not HTML Element?

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/test/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbsTest.java [hunk 3]
```diff
@@ -139,6 +139,7 @@ public void removeDeepLevelBreadcrumbsTest() {
     @Test
```
Comment (ederign) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/test/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbsTest.java]: Is there a test for:  
public void addBreadCrumb( final String associatedPerspective,
                                final String breadCrumbLabel,
                              final DefaultPlaceRequest associatedPlaceRequest,
                               final HasWidgets addTo ) {

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java [hunk 4]
```diff
@@ -197,6 +198,19 @@ public void goTo( PlaceRequest place,
...
 
+    @Override
+    public void goTo( PlaceRequest place,
+                      HTMLElement addTo ) {
+        if ( existingWorkbenchActivities.containsKey( place ) ) {
+            // if already open, behaviour is to select the place where it already lives
+            goTo( place, null, Commands.DO_NOTHING );
+        } else {
+            PanelDefinition adoptedPanel = panelManager.addCustomPanel( addTo, UnanchoredStaticWorkbenchPanelPresenter.class.getName() );
+            goTo( place, adoptedPanel, Commands.DO_NOTHING );
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Just for my curiosity, why this call is in inverse order with L195?

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManager.java [hunk 5]
```diff
@@ -132,6 +133,8 @@ PanelDefinition addWorkbenchPanel( final PanelDefinition targetPanel,
      */
     PanelDefinition addCustomPanel( HasWidgets container, String panelType );
 
+    PanelDefinition addCustomPanel( HTMLElement container, String panelType );
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManager.java]: JavaDocs.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java [hunk 6]
```diff
@@ -280,13 +289,19 @@ public void removeWorkbenchPanel( final PanelDefinition toRemove ) throws Illega
         if ( customContainer != null ) {
             customContainer.remove( presenterToRemove.getPanelView().asWidget() );
         } else {
-            final PanelDefinition parentDef = toRemove.getParent();
-            final WorkbenchPanelPresenter parentPresenter = mapPanelDefinitionToPresenter.get( parentDef );
-            if ( parentPresenter == null ) {
-                throw new IllegalArgumentException( "The given panel's parent could not be found" );
-            }
+            HTMLElement customHTMLElementContainer = customPanelsInsideHTMLElements.remove( toRemove );
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: maybe extract for a method?

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java [hunk 7]
```diff
@@ -487,6 +502,26 @@ public PanelDefinition addCustomPanel( final HasWidgets container,
...
     }
 
+    @Override
+    public PanelDefinition addCustomPanel( final HTMLElement container,
+                                           final String panelType ) {
+        PanelDefinition panelDef = new PanelDefinitionImpl( panelType );
+        final WorkbenchPanelPresenter panelPresenter = beanFactory.newWorkbenchPanel( panelDef );
+        Widget panelViewWidget = panelPresenter.getPanelView().asWidget();
+        panelViewWidget.addAttachHandler( new CustomPanelCleanupHandler( panelPresenter ) );
+        appendWidgetToElement( container, panelViewWidget );
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: Duplicated code.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 4 above)
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: also, is a duplication.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java [hunk 8]
```diff
@@ -28,6 +28,9 @@
 import com.google.gwt.user.client.ui.HasWidgets;
 import com.google.gwt.user.client.ui.IsWidget;
 import com.google.gwt.user.client.ui.Widget;
+import org.jboss.errai.common.client.dom.DOMUtil;
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: Just a question. When a perspective closes, triggers a close in all child panels. That case is also working?

**Comment by paulovmr** [comment]
jenkins retest this

File: uberfire-api/src/main/java/org/uberfire/mvp/impl/PathPlaceRequest.java (hunk 1 above)
Comment (paulovmr) [inline uberfire-api/src/main/java/org/uberfire/mvp/impl/PathPlaceRequest.java]: Yes, it has.

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbs.java (hunk 2 above)
Comment (paulovmr) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbs.java]: I maintained the previous API, which uses `HasWidgets`, and added the new Command parameter in overloaded methods. I think that if we should use `HTMLElement`, it should be dealt with in another PR.

**Comment by paulovmr** [comment]
jenkins retest this

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/test/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbsTest.java (hunk 3 above)
Comment (paulovmr) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/test/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbsTest.java]: Done.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 4 above)
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Both fixed.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java (hunk 8 above)
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: Yes, it is.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java (hunk 7 above)
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: Method extracted.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java (hunk 6 above)
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/PanelManagerImpl.java]: Done.
